README.md
# Existential Dread
A horror-themed platformer originally developed during the 2013 [Global Game Jam](https://globalgamejam.org).  

Play this project:  
https://johtull.com/existential-dread/index.html

Original 2013 Global Game Jam Entry:  
https://johtull.com/existential-dread/download/Existential+Dread+v1.2+-+GGJ+2013.jar

# Purpose
The primary purpose of this port is preservation amidst the aging technology it was originally built with (Java 1.7). Rewritten entirely in vanilla JavaScript for HTML5 canvas, we have improved the core gameplay mechanics while retaining the initial inspiration.

## History and Future
Since the Global Game Jam is a 48 hour challenge, we were unable to complete our vision for this project. Instead, we created an enjoyable gameplay experience with 3 solid levels to show off the functionality.  

With more time and motivation, I have been able to revisit this project and tailor it more to our ideal vision.  

This version has been updated to:  
- clean and modernize the codebase
- fix outstanding bugs
- add new gameplay mechanics (future)
- add new levels (future)
- add custom level editor (future)
- add new sprites (future)
- add new music (future)

### Specific Improvements
- updated tick logic for smoother/consistent gameplay
- usage of spritesheets
- use maps as hybrid configuration files instead of hardcoded values
- general cleanup

# Gameplay
Existential Dread plays like a traditional, minimal 2D platformer.

## Controls
* WASD or Arrow Keys for movement  
* SPACE or W for jumping  
* Hold SHIFT to sprint or super-jump  
* F to toggle your lantern  
  
## Stats / UI
There are 3 player stat bars in the top right corner:  
  
### Health - Top Red Bar
The amount of hit points your player has. When this reaches 0, the game is over.  
### Stamina - Middle Green Bar
The amount of sprint/super-jump energy you have. You recover stamina slowly over time.  
If you run out, you will be temporarily incapacitated while you catch your breath.  
### Battery Charge - Bottom Yellow Bar
The amount of charge your lantern has. Find more batteries to add more charge. If you run out, you will not be able to use your lantern until you find more batteries.  
Once you obtain the lantern, this bar will appear.  
  
## Objectives
Your goal is to survive the night by completing levels without succumbing to the Darkness.  
  
The moving black clouds/walls are the Darkness. For every moment in the Darkness while your 
lantern is off, you will take damage. When your health is depleted, the game is over.  

&nbsp;


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

## [1.0.4.1] - 2022-10-23
### Fixed
- Music not playing on the title screen

## [1.0.4] - 2022-10-23
### Added
- Phase II new levels
  - `lantern_hunt`

### Changed
- Added support for multiple targets on `remove` tiles
- Minor Phase II map adjustments (all maps)

### Fixed
- Stamina no longer drains when holding sprint while jumping

## [1.0.3] - 2022-10-20
### Added
- New soundtrack `fear_rising_reversed`

### Changed
- Minor Phase II map adjustments
  - `tutorial`
  - `lantern_placeholder`

### Fixed
- Up and Down arrows will no longer scroll the page
- Check for map music will no longer error when empty/undefined
- Replaced deprecated `substr` with `substring`

## [1.0.2] - 2022-10-19
### Added
- New splash image
- Phase II new levels
- Temporary variable `isGameSuccessful` for determining if a level was completed successfully

### Changed
- Replaced all `forEach` loops with `for`

### Fixed
- `updateDarkness()` now properly adds darkness on the first tick

## [1.0.1] - 2022-10-02
### Added
- Tile triggers
    - Supports buttons which remove specific tile types
    - Button tile state changes
- New levels
    - Tutorial
    - Phase II preview level
- Changelog file

### Changed
- Set background image on items to "glow"
    - render items with glow after darkness
- Added map file parameter for `resetGame()`
- Moved tile positions in sprite sheet

## [1.0.0] - 2020
### Added
- Initial port from Java to HTML5
- Sprite sheets
- Tile and Darkness properties to maps
- Base map editor configuration

js/Editor.js
EditorJS = {
    init:function() {

        document.getElementById("editor-add-tile").addEventListener("click", function() {
            let template = document.getElementById("editor-tile-template").lastElementChild.cloneNode(true);
            document.getElementById("editor-tiles").appendChild(template);
        }, false);

        document.getElementById("editor-add-darkness").addEventListener("click", function() {
            let template = document.getElementById("editor-darkness-template").lastElementChild.cloneNode(true);
            document.getElementById("editor-darknesses").appendChild(template);
        }, false);

        document.querySelector("#editor-tiles").addEventListener("change", function(e) {
            if (e.target.dataset.id === "img") {
                let selectedTile = tile_imgs[e.target.value];
                let bp = selectedTile.x * -1 + "px " + selectedTile.y * -1 + "px";
                e.target.parentNode.parentNode.querySelector(".tile-preview").style.backgroundPosition = bp;
                e.target.parentNode.parentNode.querySelector(".tile-preview").style.width = selectedTile.sizeX + "px";
                e.target.parentNode.parentNode.querySelector(".tile-preview").style.height = selectedTile.sizeY + "px";
            }
        });

        /*
        document.getElementById("editor-upload").addEventListener("change", function() {
            EditorJS.readInputFile(this.files, function(text) {
                document.getElementById("editor-text").value = text;
                //TODO fill form
            });
        }, false);

        document.getElementById("editor-download").addEventListener("click", function(e) {
            e.preventDefault();
            if(e.currentTarget.href[e.currentTarget.href.length - 1] != "#") {
                e.currentTarget.href = "#";
                e.currentTarget.textContent = "Generate";
            }else {
                EditorJS.generateDownload();
                e.currentTarget.textContent = "Download";
                document.body.appendChild(e.currentTarget);
                e.currentTarget.click();
            }
        }, false);
        */

        document.getElementById("editor-advanced-toggle").addEventListener("click", function(e) {
            let enabled = e.currentTarget.dataset.enabled;
            if(enabled == "true") {
                document.getElementById("editor-text").className = "hidden";
                document.getElementById("editor-advanced-toggle").dataset.enabled = "false";
                document.getElementById("editor-advanced-toggle").className = "red";

                document.getElementById("editor-form").className = "container block";
                document.getElementById("editor-map-object").className = "container block";
                document.getElementById("editor-tile-map").className = "container block";
                document.getElementById("editor-darkness-instructions").className = "container block";
            }else {
                document.getElementById("editor-text").className = "";
                document.getElementById("editor-advanced-toggle").dataset.enabled = "true";
                document.getElementById("editor-advanced-toggle").className = "green";

                document.getElementById("editor-form").className = "hidden";
                document.getElementById("editor-map-object").className = "hidden";
                document.getElementById("editor-tile-map").className = "hidden";
                document.getElementById("editor-darkness-instructions").className = "hidden";
            }
            
        }, false);


        // populate template image dropdown
        Object.keys(tile_imgs).forEach(function(img) {
            let option = document.createElement("option");
            option.value = img;
            option.textContent = img;
            document.getElementById("editor-tile-template").querySelector(".tile-image").appendChild(option);
        });

        // populate music dropdown
        Object.keys(music_tracks).forEach(function(track) {
            let option = document.createElement("option");
            option.value = track;
            option.textContent = track;
            document.getElementById("editor-music-dropdown").appendChild(option);
        });
    }//init
    ,
    levelEditor:function(cmd) {
        switch(cmd) {
            case "render":
                document.getElementById("title").className = "hidden";
                document.getElementById("back").className = "hidden";
                document.getElementById("gameover").className = "hidden";
                document.getElementById("editor").className = '';
                break;
            case "test":
                cancelAnimationFrame(frame);

                if(document.getElementById("editor-advanced-toggle").dataset.enabled == "false") {
                    try {
                        EditorJS.compileMap();
                    }catch(e) {
                        document.getElementById("editor-messages").textContent = "ERROR: " + e;
                        return;
                    }
                }
                document.getElementById("title").className = "hidden";
                document.getElementById("back").className = "hidden";
                document.getElementById("gameover").className = "hidden";

                resetMapPlayer();
    
                if(document.getElementById("editor-toggle-flashlight").checked) {
                    player.lanternParts = 5;
                    player.batteryCharge = 1000;
                }
    
                isEditorTesting = true;
                isLoadNextMap = false;
                isGameOver = false;
                isPaused = false;

                ctick = 0;
                ctock = 0;

                LoadMapJS.loadEditorFile();

                isStarted = true;
    
                break;
            case "clear":
                document.getElementById("editor-text").value = '';
                break;
            case "exit":
                isGameOver = true;
                isEditorTesting = false;
                renderMainTitle();
                document.getElementById("editor").className = "hidden";
                break;
        }
    }//levelEditor
    ,
    removeTile:function(element) {
        document.getElementById("editor-tiles").removeChild(element.parentNode);
    }//removeTile
    ,
    removeDarkness:function(element) {
        document.getElementById("editor-darknesses").removeChild(element.parentNode);
    }//removeDarkness
    ,
    compileMap:function() {
        let savedMap = "";
        let tiles = document.getElementById("editor-tiles").children;
        for(let i = 0; i < tiles.length; i++) {
            let row = tiles[i].children;
            let symbol = '';
            let tile = {};
            for(let j = 0; j < row.length; j++) {
                for(let k = 0; k < row[j].children.length; k++) {
                    let child = row[j].children[k];
                    if(child.nodeName === "INPUT" || child.nodeName === "SELECT") {
                        switch(child.dataset.id) {
                            case "symbol":
                                symbol = child.value;
                                break;
                            case "isSolid":
                                tile["isSolid"] = child.checked;
                                break;
                            case "sizeX":
                            case "sizeY":
                            case "offsetX":
                            case "offsetY":
                                tile[child.dataset.id] = Number(child.value);
                                break;
                            default:
                                tile[child.dataset.id] = child.value;
                                break;
                        }
                    }
                }
            }
            savedMap += symbol + "=" + JSON.stringify(tile) + "\n";
        }//for tiles

        let map = document.getElementById("editor-map-object").children;
        let mapObj = {};
        for(let i = 0; i < map.length; i++) {
            if(map[i].nodeName === "DIV") {
                let row = map[i].children;
                for(let j = 0; j < row.length; j++) {
                    if(row[j].nodeName === "INPUT" || row[j].nodeName === "SELECT") {
                        switch(row[j].dataset.id) {
                            case "width":
                            case "height":
                            case "offsetX":
                            case "offsetY":
                            case "bgImgWidth":
                            case "bgImgHeight":
                            case "batteryDrain":
                                mapObj[row[j].dataset.id] = Number(row[j].value);
                                break;
                            default:
                                mapObj[row[j].dataset.id] = row[j].value;
                                break;
                        }
                    }
                }
            }
        }//for map object        
        savedMap += "MAP=" + JSON.stringify(mapObj) + "\n";
        savedMap += "START\n"
        savedMap += document.getElementById("editor-map").value;

        document.getElementById("editor-text").value = savedMap;
    }//compileMap
    ,
    generateDownload:function() {
        let text = document.getElementById("editor-text").value;
        document.getElementById("editor-download").href = "data:text/plain;charset=utf-8," + encodeURIComponent(text);
    }//generateDownload
    ,
	readInputFile:function(fileList, callback) {
		if(typeof fileList === "undefined" || fileList.length !== 1) {
			callback("");
		}

		var fileReader = new FileReader();
		fileReader.onload = function(e) {
			callback(e.target.result);
		}
		fileReader.readAsText(fileList[0]);
	}
};

js/LoadMap.js
LoadMapJS = {
	loadTextFile:function(path, callback) {   
		var xobj = new XMLHttpRequest();
		xobj.overrideMimeType("application/json");
		xobj.open("GET", path, true);
		xobj.onreadystatechange = function () {
			if (xobj.readyState == 4 && xobj.status == "200") {
				callback(xobj.responseText);
			}
		};
		xobj.send(null);  
	}//loadTextFile
	,
	loadEditorFile:function() {
		document.getElementById("editor-messages").textContent = '';
		try {
			LoadMapJS.loadMap(document.getElementById("editor-text").value);
		}catch(e) {
			document.getElementById("editor-messages").textContent = "ERROR: " + e;
			console.log(e);
		}
	}//loadEditorFile
	,
	load32Map:function(path) {
		if(typeof path === "undefined" || path === '') {
			return;
		}
		
		LoadMapJS.loadTextFile(path, function(text) {
			LoadMapJS.loadMap(text);
		});
	}// loadMap
	,
	loadMap:function(text) {
		let tileMap = {};
		let tileList = [];
		let keyMap = {};
		let darkList = [];
		let activateList = [];
	
		let txtMap = text.replace(/\r/g, '').split('\n');
		let parseMap = false;
		let mapStartH = 0;
		
		for(var h = 0; h < txtMap.length; h++) {
			if(!parseMap) {
				if(txtMap[h] === '') {
					continue;
				}else if(txtMap[h] === "START") {
					parseMap = true;
					mapStartH = h + 1;
				}else if(txtMap[h].startsWith("DARKNESS")) {
					let dinst = txtMap[h].split('=');
					//darkList.push(Object.assign({}, JSON.parse(dinst[1].toString())));
					darkList.push(JSON.parse(dinst[1].toString()));
				}else {
					let key = txtMap[h].split('=');
					if(txtMap[h].startsWith("MAP")) {
						keyMap[key[0].toString()] = JSON.parse(key[1].toString());
					}else {
						keyMap[key[0].toString()] = JSON.parse(key[1].toString());
						
						let t = JSON.parse(key[1].toString());
						if(typeof t.sizeX === "undefined" || t.sizeX === "") {
							t.sizeX = tileRef.sizeX;
						}else {
							t.sizeX = Number(t.sizeX);
						}
						if(typeof t.sizeY === "undefined" || t.sizeY === "") {
							t.sizeY = tileRef.sizeY;
						}else {
							t.sizeY = Number(t.sizeY);
						}
						if(typeof t.isSolid === "undefined") {
							t.isSolid = tileRef.isSolid;
						}
						if(typeof t.type === "undefined") {
							t.type = tileRef.type;
						}
						if(typeof t.target !== "undefined") {
							if(!Array.isArray(t.target)) {
								t.target = [t.target]
							}else {
								t.target = t.target;
							}
						}
						if(typeof t.offsetX === "undefined") {
							t.offsetX = tileRef.offsetX;
						}else {
							t.offsetX = Number(t.offsetX);
						}
						if(typeof t.offsetY === "undefined") {
							t.offsetY = tileRef.offsetY;
						}else {
							t.offsetY = Number(t.offsetY);
						}
						if(typeof t.hasGlow === "undefined") {
							t.hasGlow = tileRef.hasGlow;
						}else {
							t.hasGlow = Boolean(t.hasGlow);
						}

						tileMap[key[0].toString()] = t;
					}
				}
			}else {
				// set level boundaries
				let width = keyMap["MAP"].width;
				//let height = keyMap["MAP"].height;
				map.floorX = keyMap["MAP"].width * tileRef.sizeX;
				map.floorY = keyMap["MAP"].height * tileRef.sizeY;

				for(var w = 0; w < width; w++) {
					let mapChar = txtMap[h].charAt(w).toString();
					if(mapChar === ' ') {
						continue;
					}else if(typeof tileMap[mapChar] !== "undefined") {
						let tmpTile = {
							"type": mapChar
						};
						if(typeof tileMap[mapChar].offsetX !== "undefined") {
							tmpTile.x = w * tileRef.sizeX + tileMap[mapChar].offsetX;
						}else {
							tmpTile.x = w * tileRef.sizeX;
						}
						if(typeof tileMap[mapChar].offsetY !== "undefined") {
							tmpTile.y = ((h - mapStartH) * tileRef.sizeY) + tileMap[mapChar].offsetY;
						}else {
							tmpTile.y = (h - mapStartH) * tileRef.sizeY;
						}

						switch(tileMap[mapChar].type) {
							case "activate":
								activateList.push(tmpTile);
								break;
							default:
								tileList.push(tmpTile);
								break;
						}
					}else if(mapChar === 'S') {
						// set spawn location
						player.x = w * player.sizeX;
						player.y = (h - mapStartH) * player.sizeY;
						// set camera offset
						map.x = Math.floor(player.x - map.screenX/2);
						if(map.x < 0) {
							map.x = 0;
						}else if(map.floorX - map.x < map.screenX) {
							map.x = map.floorX - map.screenX;
						}
						map.y = Math.floor(player.y - map.screenY/2) + player.sizeY;
						if(map.y < 0){
							map.y = 0;
						}else if(map.floorY - map.y < map.screenY) {
							map.y = map.floorY - map.screenY;
						}
					}else {
						continue;
					}
				}
			}
		}
		
		// optional camera override
		if(typeof keyMap["MAP"].cameraX !== "undefined" && keyMap["MAP"].cameraX !== "") {
			map.x = Number(keyMap["MAP"].cameraX);
		}
		if(typeof keyMap["MAP"].cameraY !== "undefined" && keyMap["MAP"].cameraY !== "") {
			map.y = Number(keyMap["MAP"].cameraY);
		}

		if(typeof keyMap["MAP"].bgImg !== "undefined" && keyMap["MAP"].bgImg != "") {
			map.background = loadImage(keyMap["MAP"].bgImg, keyMap["MAP"].bgImgWidth, keyMap["MAP"].bgImgHeight);
		}else {
			map.background = null;
		}

		if(typeof keyMap["MAP"].bgColor !== "undefined") {
			map.backgroundColor = keyMap["MAP"].bgColor;
		}else {
			map.backgroundColor = null;
		}
		
		map.nextMap = keyMap["MAP"].nextMap;
		map.tiles = tileList;
		map.tileMap = tileMap;
		map.activateList = activateList;
		map.totalMS += map.passedMS;
		map.passedMS = 0;
		map.condType = keyMap["MAP"].condType;
		map.cond = keyMap["MAP"].cond;
		if(typeof keyMap["MAP"].batteryDrain !== "undefined") {
			map.batteryDrain = keyMap["MAP"].batteryDrain;
		}else {
			map.batteryDrain = 1;
		}

		// overlay
		if(typeof keyMap["MAP"].overlayColor !== "undefined" && keyMap["MAP"].overlayColor != "") {
			map.overlayColor = keyMap["MAP"].overlayColor;
		}else {
			map.overlayColor = map.overlayAlphaReset;
		}
		if(typeof keyMap["MAP"].overlayAlpha !== "undefined" && keyMap["MAP"].overlayAlpha != "") {
			map.overlayAlpha = keyMap["MAP"].overlayAlpha;
		}else {
			map.overlayAlpha = map.overlayAlphaReset;
		}
		ctx.globalAlpha = 1;
		map.alpha = 1;

		if(debug) {
			player.laternParts = 5;
			player.batteryCharge = 1000;
		}
	
		wasd = [false, false, false, false];
		player.vertSpeed = 0;
		player.inDarkness = false;
		player.lanternOn = false;
		
		darknesses.instructions = [];
		darknesses.darkness = [];
		darknesses.instructions = darkList.reverse();
		
		if(isSoundEnabled &&
		   typeof keyMap["MAP"].music !== "undefined" &&
		   keyMap["MAP"].music !== "" &&
		   keyMap["MAP"].music !== "none") {
			loopMusic(music_tracks[keyMap["MAP"].music]);
		}else {
			stopMusic();
		}
		
		frame = requestAnimationFrame(main);
	}//loadMap
};

js/Objects.js
var debug = false;
var debugTiles = false;

// level editor
var isEditorTesting = false;

// ensure only a single animation frame is in use
var frame;

var canvas = {};
var ctx = {};

var isGameOver = false;
var isGameSuccessful = false;
var isStarted = false;
var isPaused = false;
var isSoundEnabled = true;
var isLoadNextMap = false;

var wasd = [false, false, false, false];

var tileRef = {
	x: 0,
	y: 0,
	sizeX: 32,
	sizeY: 32,
	offsetX: 0,
	offsetY: 0,
	isSolid: true,
	//isStatic: true,
	type: "",
	target: "",
	img: "",
	imgSwitch: "",
	hasGlow: false
};

function loadImage(path, x, y) {
	let img = new Image(x, y);
	img.src = path;
	return img;
}

var player_spritesheet = loadImage("img/player/original_spritesheet.png", 256, 384);
var player_imgs = {
	"idleleft": { x: 0, y: 0 },
	"idleright": { x: 128, y: 0 },
	"idlelightleft": { x: 0, y: 32 },
	"idlelightright": { x: 128, y: 32 },
	"runleft": { x: 0, y: 64 },
	"runright": { x: 128, y: 64 },
	"runlightleft": { x: 0, y: 96 },
	"runlightright": { x: 128, y: 96 },
	"exhaustedleft": { x: 0, y: 128 },
	"exhaustedright": { x: 128, y: 128 },
	"crouchleft": { x: 0, y: 160 },
	"crouchright": { x: 128, y: 160 },
	"jumpleft": { x: 0, y: 192 },
	"jumpright": { x: 128, y: 192 },
	"light": { x: 0, y: 224, sizeX: 160, sizeY: 160 }
};


var tile_spritesheet = loadImage("img/tiles/tile_spritesheet.png", 224, 128);
var tile_imgs = {
	"clear": { x: 0, y: 0 },
	"city_border": { x: 32, y: 0 },
	"city_border_half": { x: 32, y: 0, sizeY: 16 },
	"city_block": { x: 64, y: 0 },
	"city_block_half": { x: 64, y: 0, sizeY: 16},
	"factory_border": { x: 96, y: 0 },
	"factory_border_half": { x: 96, y: 0, sizeY: 16 },
	"factory_floor": { x: 128, y: 0 },
	"factory_floor_half": { x: 128, y: 0, sizeY: 16 },
	"subway_border": { x: 160, y: 0 },
	"subway_border_half": { x: 160, y: 0, sizeY: 16 },
	"subway_border_cracked": { x: 192, y: 64 },
	"subway_block": { x: 192, y: 0 },
	"subway_block_half": { x: 192, y: 0, sizeY: 16 },
	"red_door": { x: 32, y: 32 },
	"blue_door": { x: 64, y: 32 },
	"green_door": { x: 96, y: 32 },
	"original_door": { x: 0, y: 32 },
	"battery": { x: 0, y: 64 },
	"lantern_1": { x: 32, y: 64 },
	"lantern_2": { x: 64, y: 64 },
	"lantern_3": { x: 96, y: 64 },
	"lantern_4": { x: 128, y: 64 },
	"lantern_5": { x: 160, y: 64 },
	"trampoline": { x: 128, y: 48, sizeY: 16 },
	"grandma_wallpaper": { x: 160, y: 32 },
	"grandma_wallpaper_half": { x: 160, y: 48, sizeY: 16 },
	"button_off": { x: 0, y: 95, sizeY: 16 },
	"button_on": { x: 0, y: 112, sizeY: 16 },
	"glow_small": { x: 192, y: 32, sizeX: 32, sizeY: 32 },
	"glow": { x: 64, y: 96, sizeX: 160, sizeY: 160 }
};
// add default sizes if missing
Object.keys(tile_imgs).forEach(function(key) {
	if(typeof tile_imgs[key].sizeX === "undefined") {
		tile_imgs[key].sizeX = tileRef.sizeX;
	}
	if(typeof tile_imgs[key].sizeY === "undefined") {
		tile_imgs[key].sizeY = tileRef.sizeY;
	}
});

var title_imgs = {
	"splash": loadImage("img/titles/HurrrrDeerStudios2.png", 640, 480),
	"title": loadImage("img/titles/title.png", 640, 480),
	"instructions": loadImage("img/titles/instructions.png", 640, 480),
	"story": loadImage("img/titles/story.png", 640, 480),
	"credits": loadImage("img/titles/credits.png", 640, 480),
	"gameover": loadImage("img/titles/game_over.png", 640, 480),
	"victory": loadImage("img/titles/victory.png", 640, 480)
};

var music_tracks = {
	"none": "",
	"title": "audio/music/existential_dread_0.mp3",
	"hopeless": "audio/music/hopeless_1.mp3",
	"fear_rising": "audio/music/fear_rising_2.mp3",
	"fear_rising_reversed": "audio/music/fear_rising_reversed.mp3",
	"dreams_into_darkness": "audio/music/dreams_into_darkness_3.mp3"
};

var map = {
	x: 0,
	y: 0,
	floorX: 1024,
	floorY: 1024,
	screenX: 640,
	screenY: 480,
	collisionThresholdX: 10,
	clippingX: 1.01,
	gravityOn: true,
	gravitySpeed: 15,
	background: null,
	backgroundColor: null,
	tickMS: 100,
	passedMS: 0,
	totalMS: 0,
	alpha: 1,
	fadeImg: "",
	firstMapOriginal: "maps/original/map1.txt",
	firstMapPhase2: "maps/phase2/pacer.txt",
	firstMapTutorial: "maps/tutorial.txt",
	nextMap: "",
	condType: "",
	cond: "",
	tiles: [],
	tileMap: {},
	batteries: [],
	glowList: [],
	glowLightSize: 128,
	glowLightRender: 224,
	level: 0,
	overlayAlpha: 0.3,
	overlayAlphaReset: 0.3,
	overlayColor: "#000",
	overlayColorReset: "#000",
	batteryDrain: 1
};

var player = {
	img: "idleleft",
	imgTick: 0,//0-3
	imgTickMax: 3,
	isLeft: true,
	x: 64,
	y: 64,
	sizeX: 32,
	sizeY: 32,
	speed: 0,
	defaultSpeed: 5,
	defaultSprintSpeed: 7,
	defaultSprintJumpSpeed: -20,
	isSprinting: false,
	isJumping: false,
	lastX: 0,
	lastY: 0,
	vertSpeed: 0,
	statMax: 1000,
	health: 1000,
	stamina: 1000,
	sprintJumpStaminaCost: 200,
	isExhausted: false,
	exhaustedTicks: 6,
	exhaustedTicksMax: 6,
	lanternParts: 0,
	maxLanternParts: 5,
	batteryCharge: 250,
	batteryRefill: 250,
	lanternOn: false,
	inDarkness: false,
	inSafeZone: false,
	lanternLightSize: 96
};

var darkness = {
	id: 0,
	x1: 0,//left
	y1: 0,//top
	x2: 0,//right
	y2: 0,//bottom
	dir: "up", //up, down, left, right
	condType: "time", //lantern
	cond: "30", // 3
	speed: 1,
	damage: 1,
	condPosition: {
		x: 0,
		y: 0,
		handoff: true
	},
	condActivate: {
		x: 0,
		y: 0,
		type: "a",
		activate: true
	}
};

var darknesses = {
	instructions: [], // list of darkness objects loaded with the map
	darkness: [] // active darkness objects
};


function playSound(path) {
	if(isSoundEnabled && !debug) {
		sfx.pause();
		sfx.currentTime = 0;
		sfx.src = path;
		sfx.play();
	}
}
function loopSound(path) {
	if(isSoundEnabled && !debug) {
		sfx.loop = true;
		sfx.currentTime = 0;
		sfx.src = path;
		sfx.play();
	}
}
function stopSound() {
	sfx.pause();
	sfx.currentTime = 0;
}
function loopMusic(path) {
	if(isSoundEnabled && !debug) {
		music.pause();
		music.currentTime = 0;
		music.src = path;
		music.loop = true;
		music.play();
	}
}
function stopMusic() {
	music.pause();
	music.currentTime = 0;
}
function startMusic() {
	if(isSoundEnabled && !debug) {
		music.play();
		music.currentTime = 0;
	}
}
function toggleSounds() {
	isSoundEnabled = !isSoundEnabled;
	if(isSoundEnabled) {
		document.getElementById("togglesound").className = "green";
		startMusic();
	}else {
		document.getElementById("togglesound").className = "red";
		stopMusic();
	}
}

function formatRuntime(runTime) {
	let total = (runTime/1000);
	let result = "";
	if(total > 60) {
		runTimeMinutes = total / 60;
		runTimeSeconds = (runTimeMinutes % 1) * 60;
		result = ("" + Math.floor(runTimeMinutes)).padStart(2, "0") + ":" + ("" + Math.floor(runTimeSeconds)).padStart(2, "0") + (runTimeSeconds % 1).toFixed(2).substring(1);
	}else {
		result = "00:" + ("" + Math.floor(total)).padStart(2, "0") + (total % 1).toFixed(2).substring(1);
	}
	return result;
}

// sfx
var sfx = new Audio();
// TODO: fix clank.mp3 distortion
var gong = "audio/sfx/clank.wav";
var hurrrr = "audio/sfx/hurrrr.mp3";

// music
var music = new Audio();

// 100ms tick
var ctick = 0;
var ctock = 0;

// regular game tick
var tick = 60;
var tickRate = 1000/tick;
var aStartTime = 0;
var aCurrentTick = 0, aLastTick = 0, aDeltaTick = 0;
